migen/fhdl/structure.py

```diff
@@ -27,14 +27,20 @@ class _Value(DUID):
     represent the integer.
     """
     def __bool__(self):
-        # Special case: Constants and Signals are part of a set or used as
+        # Special case: Constants, Signals and Slices are part of a set or used as
         # dictionary keys, and Python needs to check for equality.
         if isinstance(self, _Operator) and self.op == "==":
             a, b = self.operands
             if isinstance(a, Constant) and isinstance(b, Constant):
                 return a.value == b.value
             if isinstance(a, Signal) and isinstance(b, Signal):
                 return a is b
+            if isinstance(a, _Slice) and isinstance(b, _Slice):
+                return (bool(a.value == b.value) and (a.start == b.start) and (a.stop == b.stop))
+            if isinstance(a, _Slice) and (isinstance(b, Constant) or isinstance(b, Signal)):
+                return (bool(a.value == b) and (a.start == 0) and (a.stop == len(b)))
+            if isinstance(b, _Slice) and (isinstance(a, Constant) or isinstance(a, Signal)):
+                return (bool(b.value == a) and (b.start == 0) and (b.stop == len(a)))
```

Comment:
I think it covers all cases. if one is a slice and the other is a Constant or a Signal. The slice can even be a slice of a slice and so on.

Reply:
>  s[:][:] == s

this should

>  s[1:2][1:2] == s[2]

this should raise a error as `s[1:2]` would have a len of 1 and for the second `[1:2]` it would need to have a len of 2.
But `s[1:2][0:1] == s[2]` should work and return `false

Reply:
I rewrote it, so it works with `s[1:3][1:2] == s[2]`.